src/dataflow/src/render/reduce.rs

```diff
@@ -606,3 +631,89 @@ fn is_min_or_max(func: &AggregateFunc) -> bool {
         _ => false,
     }
 }
+
+/// Is the aggregate function is a "max" variant.
+fn is_max(func: &AggregateFunc) -> bool {
+    match func {
+        AggregateFunc::MaxInt32
+        | AggregateFunc::MaxInt64
+        | AggregateFunc::MaxFloat32
+        | AggregateFunc::MaxFloat64
+        | AggregateFunc::MaxDecimal
+        | AggregateFunc::MaxBool
+        | AggregateFunc::MaxString
+        | AggregateFunc::MaxDate
+        | AggregateFunc::MaxTimestamp
+        | AggregateFunc::MaxTimestampTz => true,
+        _ => false,
+    }
+}
+
+/// Monoids for in-place compaction of monotonic streams.
+pub mod monoids {
+    use repr::{Datum, Row};
+    use serde::{Deserialize, Serialize};
+
+    /// A monoid containing a single-datum row, that is updated by SQL's `min`.
+    #[derive(Ord, PartialOrd, Eq, PartialEq, Debug, Clone, Serialize, Deserialize, Hash)]
+    pub struct MinMonoid {
+        pub value: Row,
+    }
+
+    use differential_dataflow::difference::Semigroup;
+    use std::ops::AddAssign;
+
+    impl<'a> AddAssign<&'a Self> for MinMonoid {
+        fn add_assign(&mut self, rhs: &'a Self) {
+            let swap = {
+                let lhs_val = self.value.unpack_first();
+                let rhs_val = rhs.value.unpack_first();
+                // Datum::Null is the identity, not a small element.
+                match (lhs_val, rhs_val) {
+                    (_, Datum::Null) => false,
+                    (Datum::Null, _) => true,
+                    (lhs, rhs) => rhs < lhs,
+                }
+            };
+            if swap {
+                self.value.clone_from(&rhs.value);
+            }
+        }
+    }
+
+    impl Semigroup for MinMonoid {
+        fn is_zero(&self) -> bool {
```

Comment:
I don't know the consequences of this implementation - imo this deserves a small comment explaining why its like this? (I'm also happy to look this up and add the comment myself after you've merged)

Reply:
I'm happy to document, but can you say more about "consequences of this implementation"? Does that entail a full description of the `Semigroup` trait?

Reply:
no -- the central confusion for me is "why is is_zero always false even though some int32 / int64s can be zero" which makes me think that zero means something else here and I don't know exactly what that is. if thats too long to write up in a comment then don't worry about it

Reply:
"Zero" in this sense is "do you promise to act as the identity under the `AddAssign` implementation?". So, for `MinMonoid` the "zero" would be positive infinity, which when added to any other element always returns the other element. For `MaxMonoid` the "zero" would be negative infinity. Neither of these are representable, so no one self-reports as zeor.

Reply:
Will add an explainer.

Reply:
but it seems like Datum::Null seems is "zero"? maybe its not worth it to do the unpacking to see if a datum is datum null here

Reply:
`Datum::Null` does something weird that violates the zero contract, which in DD at least is "is it safe to simply remove this element?". When we have a `Null` we do have to keep it, because if there are no other elements we still need to have the `Null`. In a sense, `Null` does not act as the identity on the `epsilon` empty element, in that the result is not empty but is now `Null`.